cns/dncclient/dncclient.go

```diff
@@ -0,0 +1,96 @@
+package dncclient
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"runtime"
+	"strconv"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/logger"
+	acn "github.com/Azure/azure-container-networking/common"
+)
+
+const (
+	// RegisterNodeURLFmt: /networks/{infraNetworkID}/node/{nodeID}
+	registerNodeURLFmt = "%s/networks/%s/node/%s%s"
+	// SyncNodeNetworkContainersURLFmt: /networks/{infraNetworkID}/node/{nodeID}/networkcontainers
+	syncNodeNetworkContainersURLFmt = "%s/networks/%s/node/%s/networkcontainers%s"
+
+	dncApiVersion = "?api-version=2018-03-01"
+)
+
+// NodeRegistrationRequest - Struct to hold node registration request.
+type NodeRegistrationRequest struct {
+	NumCores int `json:"NumCores"`
+}
+
+// RegisterNode registers the node with managed DNC
+func RegisterNode(httpRestService cns.HTTPService, dncEP, infraVnet, nodeID string) {
+	logger.Printf("[dncclient] Registering node: %s with Infrastructure Network: %s dncEP: %s", nodeID, infraVnet, dncEP)
+
+	var (
+		registerNodeURL = fmt.Sprintf(registerNodeURLFmt, dncEP, infraVnet, nodeID, dncApiVersion)
+		response        *http.Response
+		err             = fmt.Errorf("")
+		body            bytes.Buffer
+		httpc           = acn.GetHttpClient()
+	)
+
+	nodeRegistrationRequest := NodeRegistrationRequest{
+		NumCores: runtime.NumCPU(),
+	}
+	json.NewEncoder(&body).Encode(nodeRegistrationRequest)
+
+	for sleep := true; err != nil; sleep = true {
+		response, err = httpc.Post(registerNodeURL, "application/json", &body)
+		if err == nil {
+			if response.StatusCode == http.StatusOK {
+				var req cns.SetOrchestratorTypeRequest
+				json.NewDecoder(response.Body).Decode(&req)
+				httpRestService.SetNodeOrchestrator(&req)
+				sleep = false
+			} else {
+				err = fmt.Errorf("[dncclient] Failed to register node with http status code %s", strconv.Itoa(response.StatusCode))
+				logger.Errorf(err.Error())
+			}
+
+			response.Body.Close()
+		} else {
+			logger.Errorf("[dncclient] Failed to register node with err: %+v", err)
+		}
+
+		if sleep {
+			time.Sleep(acn.FiveSeconds)
+		}
+	}
```

Comment:
I understand that this is code that was moved, but perhaps it's a good chance to clean it up since it is a bit hard to follow. My main recommendation would be to extract the body of the loop to a function such as:
```go
func registerNode(httpCl *http.Client, url string, body io.Reader) (cns.SetOrchestratorTypeRequest, error) {
	response, err := httpCl.Post(url, acn.JsonContent, body)
	if err != nil {...}
	defer response.Body.Close()

	if response.StatusCode != http.StatusOK {...}

	var req cns.SetOrchestratorTypeRequest
	_ = json.NewDecoder(response.Body).Decode(&req)
	return req, nil
}
```
 which then makes the loop much simpler:
```go
for {
	req, err := registerNode(httpc, registerNodeURL, &body)
	if err != nil {
		logger.Errorf(err.Error())
		time.Sleep(acn.FiveSeconds)
		continue
	}
	httpRestService.SetNodeOrchestrator(&req)
	break
}
```
To go a step further, I would consider making a proper `DNCClient` struct, which embeds reusable dependencies such as an http client. The main benefit of this is that the calling code can then use an interface and stub out the actual implementation when writing tests. That can be a later refactor though. See cnsclient in the dnc repo as an example.

Reply:
Thanks, I have updated the logic. I will create the DNCClient struct as a part of next PR where I'll handle the msi in the dncclient.

Reply:
I refactored this section in #680 , can you please cross verify.